### PR TITLE
Remove todo-errors

### PR DESCRIPTION
**Pull request by FredTheDino**
Removes the todo-errors and adds in messages that are meaningful and descriptive.

### CODE REVIEW

File: sylt-compiler/src/typechecker.rs [hunk 1]
```diff
@@ -211,42 +194,34 @@ impl TypeChecker {
...
                 if let Some(_) = self.stack.iter().rfind(|v| v.ident.name == ident.name) {
                     err_type_error! {
                         self,
-                        span,
-                        ctx,
-                        todo_error!()
+                        ident.span,
+                        calling_ctx,
+                        TypeError::Exotic,
+                        "Expected a namespace with the name '{}' - but found a local variable",
```
Comment (sornas) [inline sylt-compiler/src/typechecker.rs:223]:
```suggestion
                        "Expected a namespace with the name '{}' but found a local variable",
```

File: sylt-compiler/src/typechecker.rs [hunk 2]
```diff
@@ -257,27 +232,40 @@ impl TypeChecker {
                         Some(Name::Namespace(namespace)) => Ok(TypeCtx { namespace, ..ctx }),
                         _ => err_type_error! {
                             self,
-                            span,
+                            ident.span,
                             ctx,
-                            todo_error!()
+                            TypeError::UnresolvedName(ident.name.clone()),
+                            "Failed to find namespace '{}' - did you foreget an import?",
```
Comment (sornas) [inline sylt-compiler/src/typechecker.rs:238]:
```suggestion
                            "Failed to find namespace '{}', did you forget an import?",
```

File: sylt-compiler/src/typechecker.rs [hunk 3]
```diff
@@ -257,27 +232,40 @@ impl TypeChecker {
...
                     .cloned()
                 {
                     Some(Name::Namespace(namespace)) => Ok(TypeCtx { namespace, ..ctx }),
+                    None => {
+                        return err_type_error!(
+                            self,
+                            ident.span,
+                            calling_ctx,
+                            TypeError::UnresolvedName(ident.name.clone()),
+                            "Did you foreget an import?"
```
Comment (sornas) [inline sylt-compiler/src/typechecker.rs:259]:
```suggestion
                            "Did you forget an import?"
```

File: sylt-compiler/src/typechecker.rs [hunk 4]
```diff
@@ -257,27 +232,40 @@ impl TypeChecker {
...
-                let ctx = self.type_namespace_chain(ass, ctx)?;
+                let ctx = self.type_namespace_chain(ass, calling_ctx, ctx)?;
                 match self
                     .globals
                     .get(&(ctx.namespace, ident.name.clone()))
                     .cloned()
                 {
                     Some(Name::Namespace(namespace)) => Ok(TypeCtx { namespace, ..ctx }),
+                    None => {
+                        return err_type_error!(
```
Comment (sornas) [inline sylt-compiler/src/typechecker.rs:254]:
```suggestion
                        err_type_error!(
```

Comment (sornas) [reply sylt-compiler/src/typechecker.rs:254]: untested but the other are implicit returns i think

File: sylt-compiler/src/typechecker.rs [hunk 5]
```diff
@@ -257,27 +232,40 @@ impl TypeChecker {
...
+                    }
                     _ => err_type_error! {
                         self,
-                        span,
-                        ctx,
-                        todo_error!()
+                        ident.span,
+                        calling_ctx,
+                        TypeError::Exotic,
+                        "This access should be a namespace or a blob - but it's a global {}",
```
Comment (sornas) [inline sylt-compiler/src/typechecker.rs:267]:
```suggestion
                        "This access ({}) should be a namespace or a blob but it's a global",
```

File: sylt-compiler/src/typechecker.rs [hunk 6]
```diff
@@ -287,38 +275,64 @@ impl TypeChecker {
...
                 _ => {
-                    return err_type_error!(
+                    err_type_error!(
                         self,
                         ident.span,
-                        ctx,
+                        calling_ctx,
                         TypeError::Exotic,
-                        "Cannot find type '{}' - is it perhaps a type-variable?",
+                        "Expected a blob - '{}' should be a blob",
```
Comment (sornas) [inline sylt-compiler/src/typechecker.rs:303]:
```suggestion
                        "Expected a blob but got '{}'",
```

File: sylt-compiler/src/typechecker.rs [hunk 7]
```diff
@@ -287,38 +275,64 @@ impl TypeChecker {
...
+                            "Expected a blob"
+                        )
+                    }
+                    _ => {
+                        err_type_error!(
+                            self,
+                            ident.span,
+                            calling_ctx,
+                            TypeError::Exotic,
+                            "Expected a blob - '{}' should be a blob",
```
Comment (sornas) [inline sylt-compiler/src/typechecker.rs:332]:
```suggestion
                            "Expected a blob but got '{}'",
```

File: sylt-compiler/src/typechecker.rs [hunk 8]
```diff
@@ -439,7 +453,15 @@ impl TypeChecker {
...
                         // NOTE(ed): This disallowes type-variables that are only used for
                         // constraints.
-                        None => return err_type_error!(self, span, ctx, todo_error!()),
+                        None => {
+                            return err_type_error!(
+                                self,
+                                span,
+                                ctx,
+                                TypeError::UnresolvedName(var.clone()),
+                                "Unused type-variable (only usages in the function signature are counted)"
```
Comment (sornas) [inline sylt-compiler/src/typechecker.rs:462]:
```suggestion
                                "Unused type-variable. (Only usages in the function signature are counted)"
```

Comment (sornas) [reply sylt-compiler/src/typechecker.rs:462]: but i'm not sure how to handle multiple sentences when we don't usually include a period.

Comment (sornas) [reply sylt-compiler/src/typechecker.rs:462]: maybe we should be able to write multiple lines? and let the error reporting handle indentation.

so we could do
```
Unusued type variable
(Only usages in the function signature are counted)
```

File: sylt-compiler/src/typechecker.rs [hunk 9]
```diff
@@ -1832,7 +1854,14 @@ impl TypeChecker {
                     self.unify(span, ctx, kx, ys[0])?;
                     self.unify(span, ctx, vx, ys[1]).map(|_| ())
                 } else {
-                    err_type_error!(self, span, ctx, todo_error!())
+                    err_type_error!(
+                        self,
+                        span,
+                        ctx,
+                        TypeError::Violating(self.bake_type(b)),
+                        "Expected a tuple of length 2 - but got length {}",
```
Comment (sornas) [inline sylt-compiler/src/typechecker.rs:1862]:
```suggestion
                        "Expected a tuple of length 2 but got {}",
```

File: sylt-compiler/src/typechecker.rs (hunk 2 above)
Comment (sornas) [reply sylt-compiler/src/typechecker.rs:238]: matches with error below

File: sylt-compiler/src/typechecker.rs [hunk 10]
```diff
@@ -176,30 +170,19 @@ impl TypeChecker {
         ty_id
     }
 
-    fn namespace_chain(&self, assignable: &Assignable, ctx: TypeCtx) -> TypeResult<TypeCtx> {
-        let span = assignable.span;
+    fn namespace_chain(&self, assignable: &Assignable, ctx: TypeCtx) -> Option<TypeCtx> {
```
Comment (EMattfolk) [inline sylt-compiler/src/typechecker.rs:173]: Wait, I've seen this one before.

**Review by EMattfolk — APPROVED** [review]
I feel like there are more new errors than there are new tests. Wasn't the point of the todo-error that we fix it when we get to it? Looks good either way.

File: sylt-compiler/src/typechecker.rs (hunk 10 above)
Comment (FredTheDino) [reply sylt-compiler/src/typechecker.rs:173]: Yeah - I sometimes realize I'm wrong ;)

File: sylt-compiler/src/typechecker.rs (hunk 8 above)
Comment (FredTheDino) [reply sylt-compiler/src/typechecker.rs:462]: A great PR - feel free to do it!

**Review by FredTheDino — COMMENTED** [review]
Fix typos and style

**Comment by FredTheDino** [comment]
> I feel like there are more new errors than there are new tests. Wasn't the point of the todo-error that we fix it when we get to it? Looks good either way.

Got sick of writing tests for single error messages - it feels like really poorly spent time

**Pull request merged**